### PR TITLE
Nebm refactoring

### PR DESCRIPTION
**Pull request by davidcortesortuno**
Refactoring of the old NEB method code. I added the Geodesic distances as in Bessarab et al, paper (Computer Physics Communications 196 (2015) 335-347). I also added tests for the library and a couple of examples. The NEB codes are divided in 4 files:

**nebm_base**: The base class with common methods

**nebm_spherical**: Using spherical coordinates for the magnetisation and Euclidean distances. Tangents are not projected. The evolution of the method is based on Dittrich et al paper (Dittrich et al., JMMM 250 (2002) L12-L19). However, spherical angles here are not well defined, so this code does not work for some systems (in an email, Bessarab mentioned that Dittrich et al paper had some problems)

**nebm_cartesian**: Using Cartesian coordinates for the magnetisation and Euclidean distances. Tangents are not projected. The evolution is based on Suess et al. paper using a LLG equation but adding an extra term to fix the magnetisation length. According to the tests, it seems the Euclidean distance is not optimal for the method. It works ok for most of the systems but it seems that for large skyrmion systems there is a lack of proper convergence due to the spring force. For basic examples it converges nicely.

**nebm_geodesic**: Completely based on Bessarab's paper, using Cartesian coordinates for the magnetisation and Geodesic distances. Tangents and effective force are projected, making the method to converge faster towards a minimum and avoid issues with the spring force (we can set `k` in a wide range and always find convergence). The evolution equation is the same than in the Cartesian method. 

To consider:

* Since Euclidean distances are not the optimal ones, we could maybe remove the Spherical and Cartesian codes in the future and rather use Geodesic-Cartesian and Geodesic-Spherical (this one would require another integrator, like a velocity projection). The class names are maybe a bit misleading, because the main difference of the Cartesian code with the Geodesic code is the definition of the distances. 

* Most of the functions are in the C / Cython libraries which I separated according to the coordinate system. The relevant ones are in the `base` and `Cartesian` codes. I also set up the Climbing Image technique, which can be used with every of the 3 codes.

* I still need to write a proper documentation of the code to show how the files are structured and how the code could be extended.

### CODE REVIEW

**Comment by ryanpepper** [comment]
Hi David, I'll have a look when I'm back on Thursday :)

**Comment by fangohr** [comment]
Thanks David, thanks Ryan.

File: fidimag/common/nebm_tools.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+from __future__ import print_function
+from __future__ import division
+import numpy as np
+
+
+def cartesian2spherical(y_cartesian):
```
Comment (ryanpepper) [inline fidimag/common/nebm_tools.py]: Maybe add a comment just to point out which definition of spherical coords you've used, i.e what are the bounds of theta and phi

File: fidimag/common/nebm_tools.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    difference to be smaller than PI
+
+    We scale the norm by the array size
+
+    """
+
+    y = np.copy(A)
+
+    if scale:
+        y = np.sqrt(np.sum(y ** 2.)) / len(y)
```
Comment (ryanpepper) [inline fidimag/common/nebm_tools.py]: I think there's a function called np.linalg.norm that does this

File: fidimag/common/nebm_tools.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    #          [0] ]                    [1  3  4  5] ]
+    interpolations = np.zeros((n, 1))
+    interpolations = interpolations + y_initial
+
+    # We will not interpolate pinned spins
+    if pins is None:
+        #  Just use half the length of y_initial
+        pins = np.zeros(len(y_initial[::2]))
+
+    # Since we have a pin index per every PAIR of coordinates, we copy very
```
Comment (ryanpepper) [inline fidimag/common/nebm_tools.py]: every entry

File: fidimag/common/nebm_tools.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    #          [0] ]                    [1  3  4  5] ]
+    interpolations = np.zeros((n, 1))
+    interpolations = interpolations + y_initial
+
+    # We will not interpolate pinned spins ------------------------------------
+    if pins is None:
+        # Only use 1/3 of the length of y_initial (1 pin per mesh/lattice site)
+        pins = np.zeros(len(y_initial[::3]))
+
+    # Since we have a pin index per every TRIAD of coordinates, we copy very
```
Comment (ryanpepper) [inline fidimag/common/nebm_tools.py]: every

File: fidimag/common/nebm_tools.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    # -------------------------------------------------------------------------
+
+    # We will perform the calculations for every spin at once
+    y_initial.shape = (-1, 3)
+    y_final.shape = (-1, 3)
+
+    # The cross products of corresponding spins in the initial and final images
+    yi_cross_yf = np.cross(y_initial, y_final)
+    # This should only be an array of ones:
+    yi_cross_yf_norm = np.apply_along_axis(lambda x: np.sqrt(np.sum(x ** 2)),
```
Comment (ryanpepper) [inline fidimag/common/nebm_tools.py]: np.linalg.norm has an axis argument so you don't need to do this

**Comment by ww1g11** [comment]
It's very cool to introduce the Geodesic distances, @davidcortesortuno

**Comment by ww1g11** [comment]
It's very clever to use the Cython types rather than the default numpy declarations, we probably need to do the same thing for the whole fidimag. @davidcortesortuno

**Pull request merged**